### PR TITLE
Migrate deprecated Microsoft.Azure.Storage.Blob to  Azure.Storage.Blobs

### PR DESCRIPTION
**Pull request by djsmith85**
## Type of change
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective
Migrate from the deprecated `Microsoft.Azure.Storage.Blob` (V11 api) to `Azure.Storage.Blobs` (V12 api)
Migrate from package `Microsoft.AspNetCore.DataProtection.AzureStorage` to `Azure.Extensions.AspNetCore.DataProtection.Blobs`

Guide used for migration: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md

Biggest change with the V12 API is that MetaData and HttpHeaders have to be retrieved separately and sent fully when uploading/updating a blob or else the data gets wiped only new values are persisted.

Using our test storage account, I verified that the change in dataprotection still pulls the dataprotection.pfx from storage and also persist keys in the `aspnet-dataprotection`-container within the blob `keys.xml`

Asana task for DataProtection: https://app.asana.com/0/1200804338582616/1201408153961033/f
Asana task for Azure.Storage.Blobs: https://app.asana.com/0/1200804338582616/1201301984263844/f

## Code changes
* **src/Core/Core.csproj:** 
    * Remove `Microsoft.Azure.Storage.Blob` and install `Azure.Storage.Blobs`
    * Remove `Microsoft.AspNetCore.DataProtection.AzureStorage` and install `Azure.Extensions.AspNetCore.DataProtection.Blobs`
* **src/Core/Utilities/CoreHelpers.cs:**
    * Update `usings` to use `Azure.Storage.Blobs` instead of `Microsoft.Azure.Storage.Blob`
    *  Order `usings`
    * `GetBlobCertificateAsync`
        * Pass in storageConnectionString (globalSettings) instead of CloudStorageAccount
        * Use new `BlobServiceClient` as a replacement for `CloudStorageAccount.CreateCloudBlobClient`
        * Reduce costs/api-calls by not calling `ExistsAsync`. If retrieving a container or blob fails, we'll catch the error and return null
    * `GetIdentityServerCertificate`
        * Remove creation of `storageAccount` and update call to `GetBlobCertificateAsync` changed method signature
* **src/Core/Utilities/ServiceCollectionExtensions.cs:**
    * Using the DataProtection provided by `Azure.Extensions.AspNetCore.DataProtection.Blobs` instead of Microsoft.AspNetCore.DataProtection.AzureStorage to set AppName and persist keys to the Azure Blob Storage
    * Remove creation of `storageAccount` to retrieve `dataprotection.pfx`
* **src/Core/Services/Implementations/LicensingService.cs:** 
    * Remove creation of `storageAccount` to retrieve `licensing.pfx`

* **src/Core/Services/Implementations/AzureAttachmentService.cs: and** 
* **src/Core/Services/Implementations/AzureSendFileStorageService.cs:** 
    * Update `usings` to use `Azure.Storage.Blobs` instead of `Microsoft.Azure.Storage.Blob`
    * Migrate usage to `Azure.Storage.Blobs` 
     * Reduce costs/api-calls by not calling `ExistsAsync`. If retrieving a `BlobProperties` via `GetPropertiesAsync` fails, we'll catch the error and return

## Testing requirements
Regression testing on CipherAttachments and Send functionality
Loading IdentityCertificate from AzureStorageAccount (`container = 'certificates', blobName='identity.pfx'`) 
Loading DataProtection Certificate from AzureStorageAccount (`container = 'certificates', blobName='dataprotection.pfx'`) and saving kets to AzureBlobStorage (`container = 'aspnet-dataprotection', blobName='keys.xml'`)

## Before you submit
- [ ] If making database changes - I have also updated Entity Framework queries and/or migrations
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

File: src/Core/Utilities/CoreHelpers.cs [hunk 1]
```diff
@@ -253,20 +252,22 @@ public static string GetEmbeddedResourceContentsAsync(string file)
...
-                {
-                    var blobBytes = new byte[blobRef.Properties.Length];
-                    await blobRef.DownloadToByteArrayAsync(blobBytes, 0).ConfigureAwait(false);
-                    return new X509Certificate2(blobBytes, password);
-                }
+                var blobServiceClient = new BlobServiceClient(connectionString);
+                var containerRef2 = blobServiceClient.GetBlobContainerClient(container);
+                var blobRef = containerRef2.GetBlobClient(file);
+
+                using var memStream = new MemoryStream();
```
Comment (MGibson1) [inline src/Core/Utilities/CoreHelpers.cs:263]: I'm shaky on this syntax, when is this memstream disposed of? line 265?

Comment (MGibson1) [reply src/Core/Utilities/CoreHelpers.cs:263]: What happens on exception?

File: src/Core/Services/Implementations/AzureAttachmentStorageService.cs [hunk 2]
```diff
@@ -49,121 +56,130 @@ public class AzureAttachmentStorageService : IAttachmentStorageService
...
         }
 
         public async Task<string> GetAttachmentUploadUrlAsync(Cipher cipher, CipherAttachment.MetaData attachmentData)
         {
             await InitAsync(EventGridEnabledContainerName);
-            var blob = _attachmentContainers[EventGridEnabledContainerName].GetBlockBlobReference(BlobName(cipher.Id, attachmentData));
+            var blobClient = _attachmentContainers[EventGridEnabledContainerName].GetBlobClient(BlobName(cipher.Id, attachmentData));
             attachmentData.ContainerName = EventGridEnabledContainerName;
-            var accessPolicy = new SharedAccessBlobPolicy()
+            if (blobClient.CanGenerateSasUri)
```
[the remark this answers is not in the document]
Comment (djsmith85) [reply src/Core/Services/Implementations/AzureAttachmentStorageService.cs:78]: If I'm not mistaking it essentially just checks if the Storage Account was authenticated with a StorageSharedKeyCredential. Which we do using the `_globalSettings.Attachment.ConnectionString` when creating the `BlobServiceClient` from which we initialize `BlobContainerClient`s and then the actual `BlobClient`. In the previous code we did not check for this, and as we are always using the same connectionString and Storage Account, we could probably also remove the check.

[the remark this answers is not in the document]
Comment (MGibson1) [reply src/Core/Services/Implementations/AzureAttachmentStorageService.cs:78]: I'm not sure the check is bad, per se, I'm just worried we're not already checking for a null string returned here. I think that's changing behaviors

**Review by MGibson1 — COMMENTED** [review]
Sorry I've been slacking on this PR.

Just the one concern left

File: src/Core/Services/Implementations/AzureAttachmentStorageService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (djsmith85) [reply src/Core/Services/Implementations/AzureAttachmentStorageService.cs:78]: You are right, we currently do not check for null. As the check would have to happen on the receiving end, it's better to bubble the exception if anything throws, which I'm pretty certain it never will (due to the storage account set). We'd notice very fast if the storage account in prod was changed and SAS tokens were invalid.

Keeping the CanGenerateSasUri does change behavior

**Review by MGibson1 — APPROVED** [review]

**Pull request merged**